Tighten up error checking on manual_kernel_registration

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49505 Add at::cpu namespace of functions for structured kernels
* #49498 Push anonymous namespace into codegen, not template
* #49490 Make use_c10_dispatcher: full mandatory for structured kernels
* #49489 Move default or no default logic into native.argument
* #49348 codegen: Resolve overload ambiguities created by defaulted arguments
* **#49341 Tighten up error checking on manual_kernel_registration**
* #49245 Construct CppSignatureGroup from NativeFunction

I noticed that #49097 was using manual_kernel_registration incorrectly,
so this diff tightens up the testing so that:

1. We don't generate useless wrapper functions when manual_kernel_registration
is on (it's not going to be registered, so it does nothing).

2. manual_kernel_registration shouldn't affect generation of functions in
Functions.h; if you need to stop bindings, use manual_cpp_binding

3. Structured and manual_kernel_registration are a hard error

4. We raise an error if you set dispatch and manual_kernel_registration at the
same time.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25594003](https://our.internmc.facebook.com/intern/diff/D25594003)

ezyang committed:
Tighten up error checking on manual_kernel_registration

I noticed that #49097 was using manual_kernel_registration incorrectly,
so this diff tightens up the testing so that:

1. We don't generate useless wrapper functions when manual_kernel_registration
is on (it's not going to be registered, so it does nothing).

2. manual_kernel_registration shouldn't affect generation of functions in
Functions.h; if you need to stop bindings, use manual_cpp_binding

3. Structured and manual_kernel_registration are a hard error

4. We raise an error if you set dispatch and manual_kernel_registration at the
same time.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

albanD commented:
I am not sure to understand how this change interacts with this check: https://github.com/pytorch/pytorch/blob/25833e5d1cba1ba3a8631d4014c37879337d00dc/tools/autograd/gen_variable_type.py#L380-L385

Does this mean that if you set manual_kernel_registration, you cannot put an entry in derivatives.yaml (for it to generate Function for you for example)?

ezyang commented:
@albanD This case seems pretty logical to me: if you are manually writing the Autograd implementation, you better not put anything in derivatives.yaml, because it's going to be ignored!

albanD commented:
But what if I want to re-use the backward Node that gets generated and I only want to manually write the VariableType kernel?

ezyang commented:
> But what if I want to re-use the backward Node that gets generated and I only want to manually write the VariableType kernel?

This seems like a weird thing to want to do. In particular, in this sort of arrangement, you're relying on derivatives.yaml to *implicitly* specify the captures you want, but the code itself doesn't actually do anything and gets ignored by the real implementation. That's gotta be hella confusing for a potential reader.

But, stepping back a moment, in general, I would expect part of one's responsibilities as autograd maintainer to include refactoring and updating code to let you do things like this. For example, when we were in a similar situation with the regular codegen, where we wanted Python codegen but not C++ codegen, we simply [added the knob](https://github.com/pytorch/pytorch/pull/49092) necessary to do control in this way. If it is hard to make changes like that, well, maybe that's a good sign we should get @ljk53 to finish rewriting the autograd codegen to have static types ;)

albanD commented:
> This seems like a weird thing to want to do.

It just happened to be the exact thing I needed in this case: let the codegen do all the things I can re-use and just change the one I want to change.
In my mind a couple functions in VariableTypeManual were based on copy pasting the codegen result and adapted. But looking into it in more details, it looks like it has been cleaned up quite a lot. I will do something else in this case.


And yes we really need to do that update to the codegen! :)

ezyang commented:
> It just happened to be the exact thing I needed in this case: let the codegen do all the things I can re-use and just change the one I want to change.

Just to be clear, it does seem like a legitimate use case. But it has implications (e.g., now the way we name local bindings in codegen is no longer a private implementation detail, as manually written code relies on the naming) and it is better to be explicit about what a hack the current implementation is. I want to be flexible about letting PRs in even if they are not perfect, but part of the two way street here is being clear about when things are hacks and agreeing what the correct structure should be when it can't be conveniently be done in the short term.

albanD commented on tools/codegen/gen.py:
```diff
@@ -481,6 +482,8 @@ def gen_unstructured(self, f: NativeFunction) -> Optional[str]:
 
         if self.dispatch_key not in f.dispatch:
             return None
+        if f.manual_kernel_registration:
```
Is this change actually changing what gets generated or not? From what I can see, it is not equivalent to the removal from `ComputeFunction` below right?

ezyang replied:
Yes, I am generating less code. Fortunately the code was all dead so there is no visible impact.

albanD reviewed (commented):
I think the description of the change makes sense. It took me a bit of time to understand it.
This new error will definitely be helpful for people adding new functions!

albanD reviewed (approved):
Reducing compilation time!

ezyang committed:
Update on "Tighten up error checking on manual_kernel_registration"


I noticed that #49097 was using manual_kernel_registration incorrectly,
so this diff tightens up the testing so that:

1. We don't generate useless wrapper functions when manual_kernel_registration
is on (it's not going to be registered, so it does nothing).

2. manual_kernel_registration shouldn't affect generation of functions in
Functions.h; if you need to stop bindings, use manual_cpp_binding

3. Structured and manual_kernel_registration are a hard error

4. We raise an error if you set dispatch and manual_kernel_registration at the
same time.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Tighten up error checking on manual_kernel_registration"


I noticed that #49097 was using manual_kernel_registration incorrectly,
so this diff tightens up the testing so that:

1. We don't generate useless wrapper functions when manual_kernel_registration
is on (it's not going to be registered, so it does nothing).

2. manual_kernel_registration shouldn't affect generation of functions in
Functions.h; if you need to stop bindings, use manual_cpp_binding

3. Structured and manual_kernel_registration are a hard error

4. We raise an error if you set dispatch and manual_kernel_registration at the
same time.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25594003](https://our.internmc.facebook.com/intern/diff/D25594003)

[ghstack-poisoned]

ezyang committed:
Update on "Tighten up error checking on manual_kernel_registration"


I noticed that #49097 was using manual_kernel_registration incorrectly,
so this diff tightens up the testing so that:

1. We don't generate useless wrapper functions when manual_kernel_registration
is on (it's not going to be registered, so it does nothing).

2. manual_kernel_registration shouldn't affect generation of functions in
Functions.h; if you need to stop bindings, use manual_cpp_binding

3. Structured and manual_kernel_registration are a hard error

4. We raise an error if you set dispatch and manual_kernel_registration at the
same time.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25594003](https://our.internmc.facebook.com/intern/diff/D25594003)

[ghstack-poisoned]

ezyang committed:
Update on "Tighten up error checking on manual_kernel_registration"


I noticed that #49097 was using manual_kernel_registration incorrectly,
so this diff tightens up the testing so that:

1. We don't generate useless wrapper functions when manual_kernel_registration
is on (it's not going to be registered, so it does nothing).

2. manual_kernel_registration shouldn't affect generation of functions in
Functions.h; if you need to stop bindings, use manual_cpp_binding

3. Structured and manual_kernel_registration are a hard error

4. We raise an error if you set dispatch and manual_kernel_registration at the
same time.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25594003](https://our.internmc.facebook.com/intern/diff/D25594003)

[ghstack-poisoned]
